ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift

```diff
@@ -0,0 +1,440 @@
+
+    navigationController?.toolbar?.barTintColor = UIColor.orange
+
+    self.configureContentView()
+    self.configureDefaultHeightButton()
+    self.configureLabel()
+    self.configureKeyboardImage()
+    self.updateKeyboardImage()
+    self.configureKeyboardConstraints()
+    self.configureKeyboardResizer()
+  }
+  
+  override func viewDidAppear(_ animated: Bool) {
+    super.viewDidAppear(animated)
+ 
+    self.calculateKeyboardHeightLimits()
+  }
+  
+  private func determineOrientation() {
+    self.isPortrait = UIScreen.main.bounds.height > UIScreen.main.bounds.width
+    let message = "determineOrientation, isPortrait: \(self.isPortrait)"
+    os_log("%{public}s", log:KeymanEngineLogger.ui, type: .debug, message)
+  }
+  
+  private func determineDefaultKeyboardHeights() {
+    // if no KeyboardScaleMap found for device, then default to 216.0
+    let portraitKeyboardScale = KeyboardScaleMap.getDeviceDefaultKeyboardScale(forPortrait: true)
+    self.defaultPortraitHeight = Double(portraitKeyboardScale?.keyboardHeight ?? 216.0)
+
+    let landscapeKeyboardScale = KeyboardScaleMap.getDeviceDefaultKeyboardScale(forPortrait: false)
+    self.defaultLandscapeHeight = Double(landscapeKeyboardScale?.keyboardHeight ?? 216.0)
+  }
+  
+  /**
+   * Read the previously set keyboard heights from UserDefaults.
+   * Running the first time, these values would be defaults set in KeymanWebviewController.initKeyboardSize()
+   * These values should be set prior to now; if not, log an error and use  default values
+   */
+  private func applyKeyboardHeight() {
+    if (self.isPortrait) {
+      if (Storage.active.userDefaults.object(forKey: Key.portraitKeyboardHeight) != nil) {
+        self.keyboardHeight = Storage.active.userDefaults.portraitKeyboardHeight
+        let message = "applyKeyboardHeight, from UserDefaults loaded portrait value \(self.keyboardHeight)"
+        os_log("%{public}s", log:KeymanEngineLogger.ui, type: .debug, message)
+      } else {
+        self.keyboardHeight = self.defaultPortraitHeight
+        let message = "applyKeyboardHeight, portraitHeight not found in UserDefaults, using default value \(self.keyboardHeight)"
+        os_log("%{public}s", log:KeymanEngineLogger.ui, type: .error, message)
+      }
+    } else {
+      if (Storage.active.userDefaults.object(forKey: Key.portraitKeyboardHeight) != nil) {
+        self.keyboardHeight = Storage.active.userDefaults.landscapeKeyboardHeight
+        let message = "applyKeyboardHeight, from UserDefaults loaded landscape value \(self.keyboardHeight)"
+        os_log("%{public}s", log:KeymanEngineLogger.ui, type: .debug, message)
+      } else {
+        self.keyboardHeight = self.defaultLandscapeHeight
+        let message = "applyKeyboardHeight, landscapeHeight not found in UserDefaults, using default value \(self.keyboardHeight)"
+        os_log("%{public}s", log:KeymanEngineLogger.ui, type: .error, message)
+      }
+    }
```

Comment:
There's a fair bit of repetition between landscape and portrait orientation code. Could we merge the values into an array (0=portait, 1=landscape) or something like that? What do you think? Or an {x,y}?

Reply:
I tried to make the new code just use a keyboard height where I could and be unaware of orientation.  For the code that needs to be aware of orientation, I don't see any real advantage over using individual named fields. I think we can assume there will only be two, and we don't have mentally map to some other structure when reading or writing the values based on the current orientation.